admin_guide/out_of_resource_handling.adoc

```diff
@@ -0,0 +1,270 @@
+starved compute resource relative to the pods scheduling request.
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first.
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+A `Guaranteed` pod is guaranteed to never be evicted because of another pod's
+resource consumption.  If a system daemon (i.e. node, `docker`, `journald`, etc.)
+is consuming more resources than were reserved via `system-reserved` or `kube-reserved` allocations,
+and the node only has `Guaranteed` pod(s) remaining, then the node must choose to evict a
+`Guaranteed` pod in order to preserve node stability, and to limit the impact
+of the unexpected consumption to other `Guaranteed` pod(s).
+
+=== Scheduler
+
+The node will report a condition when a compute resource is under pressure.  The
+scheduler views that condition as a signal to dissuade placing additional
+pods on the node.
+
+.Node Conditions and Scheduler Behavior
+[cols="3a,8a",options="header"]
+|===
+
+|Node Condition |Scheduler Behavior
+
+|`*MemoryPressure*`
+| `BestEffort` pods are not scheduled to the node.
+|===
+
+== Node OOM Behavior
+
+If the node experiences a system OOM (out of memory) event prior to the node is able to reclaim memory,
+the node depends on the [oom_killer](https://lwn.net/Articles/391222/) to respond.
+
+The node sets a `oom_score_adj` value for each container based on the quality of service for the pod.
+
+.Quality of Service OOM Scores
+[cols="3a,8a",options="header"]
+|===
+
+| Quality of Service |oom_score_adj
+
+| `Guaranteed`
+| -998
+
+| `BestEffort`
+| 1000
+
+| `Burstable`
+| min(max(2, 1000 - (1000 * memoryRequestBytes) / machineMemoryCapacityBytes), 999)
+|===
+
+If the node is unable to reclaim memory prior to a node experiencing system OOM, the `oom_killer` will calculate
+an `oom_score` based on the percentage of memory its using on the node, and then add the `oom_score_adj` to get an
+effective `oom_score` for the container, and then kills the container with the highest score.
+
+The intended behavior should be that containers with the lowest quality of service that
+are consuming the largest amount of memory relative to the scheduling request should be killed first in order
+to reclaim memory.
+
+Unlike pod eviction, if a pod container is OOM killed, it may be restarted by the node based on its `RestartPolicy`.
+
+== Best Practices
+
+=== Disable swap memory
+
+Failure to disable swap memory will make the node not recognize it is under *MemoryPressure*.
```

Comment:
To be clear, swap was never supported in upstream Kubernetes.  Quality of service features have never been swap aware.  The issue now is that there are fundamental node reliability features if used in conjunction with swap will not function that now allow a cluster to better auto-balance memory consumption by evicting pods for placement on other nodes, and detecting memory pressure to prevent scheduling.

I never agreed to the change that stated that swap is not recommended to be disabled.  It was made in https://github.com/openshift/openshift-docs/pull/2094 , and I think that change was made in error.

As for making cAdvisor swap aware, the primary issue is where tmpfs backed volumes are accounted.  It's possible it could improve when we have pod level cgroups.  See https://github.com/google/cadvisor/pull/1380